### PR TITLE
Avoid registering same gossip twice

### PR DESCRIPTION
**Pull request by jml**
Fixes crasher bug in alertmanager.

During local testing, I came across a condition where we could crash the alertmanager.  What would happen is that we'd create a `NewAlertmanager`, which creates a `nflog`, which creates a `mesh.Gossip` on the `Router`.

If that alertmanager creation process failed for any reason, then we'd abandon it. 

However, when we tried to create that alertmanager again (i.e. once every config poll loop), then we'd create the `nflog` again, which tries to create a `mesh.Gossip` on the same `Router`, but fails, due to https://github.com/weaveworks/mesh/blob/2534f734d9a18501c3b155cd24877fca0696fb18/router.go#L136 (see also https://github.com/weaveworks/mesh/issues/69)

To work around this, I've created a `gossipFactory` class which keeps track of which IDs have been used.

In the process, while figuring out how the code worked, I did a few small refactorings:
* extracted `setConfig` out of `addNewConfigs`, to have a method that operates on a single config
* extracted `newAlertmanager` out of `setConfig`, to help me distinguish between the book-keeping for `MultitenantAlertmanager` and the process of creating a new, valid `Alertmanager`.
* created a new `Router` interface, to give me some idea of how we're using `mesh.Router`

### CODE REVIEW

File: alertmanager/router.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package alertmanager
+
+import (
+	"sync"
+
+	"github.com/weaveworks/mesh"
+)
+
+// Router is the interface we use for a mesh router.
```
Comment (juliusv) [inline alertmanager/router.go]: `Router` sounds so generic and could in Alertmanager terms easily be confused with alert routing. Maybe `GossipRouter`?

Comment (juliusv) [inline alertmanager/router.go]: I'd probably also call the file gossip.go too, but then we already have mesh.go, and now I'm wondering if the two files should be combined.

Comment (jml) [inline alertmanager/router.go]: I have no good intuition for how Go files within a package should be organised, so whatever you think is best.

File: alertmanager/router.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package alertmanager
+
+import (
+	"sync"
+
+	"github.com/weaveworks/mesh"
+)
+
+// GossipRouter is the interface we use for a mesh router.
+type GossipRouter interface {
```
Comment (juliusv) [inline alertmanager/router.go]: Forgot: this type and its methods are only used within this package, so un-export them.

File: alertmanager/router.go (hunk 1 above)
Comment (juliusv) [inline alertmanager/router.go]: Yeah, I think it makes sense to put it in one file, as it's both related to mesh and routing.

**Comment by juliusv** [comment]
:+1: otherwise

**Pull request merged**